File: server/src/main/java/io/iamcyw/ams/ResourceManager.java
package io.iamcyw.ams;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.iamcyw.tower.commandhandling.gateway.CommandGateway;
import io.iamcyw.tower.quarkus.runtime.DomainNameMappings;
import io.iamcyw.tower.queryhandling.gateway.QueryGateway;
import io.quarkus.vertx.web.Route;
import io.quarkus.vertx.web.RoutingExchange;
import io.smallrye.common.annotation.Blocking;

import javax.enterprise.context.ApplicationScoped;
import java.util.List;

@ApplicationScoped
public class ResourceManager {
    private final CommandGateway commandGateway;

    private final QueryGateway queryGateway;

    private final DomainNameMappings registry;

    private final ObjectMapper objectMapper;

    public ResourceManager(CommandGateway commandGateway, QueryGateway queryGateway, DomainNameMappings registry,
                           ObjectMapper objectMapper) {
        this.commandGateway = commandGateway;
        this.queryGateway = queryGateway;
        this.registry = registry;
        this.objectMapper = objectMapper;
    }

    @Route(path = "command/send")
    @Blocking
    public void send(RoutingExchange ex) {
        JsonRpcRequest request = ex.context().getBodyAsJson().mapTo(JsonRpcRequest.class);
        try {
            if (request.params().size() == 1) {
                commandGateway.send(codec(request.params().get(0), getDomain(request.method())));
            }
            ex.ok().end();
        } catch (Exception e) {
            ex.serverError().send(exception(e, request));
        }
    }

    @Route(path = "command/request")
    @Blocking
    public void request(RoutingExchange ex) {
        JsonRpcRequest request = ex.context().getBodyAsJson().mapTo(JsonRpcRequest.class);
        try {
            Object result = commandGateway.request(codec(request.params().get(0), getDomain(request.method())));
            ex.ok().end(success(result, request));
        } catch (Exception e) {
            ex.serverError().send(exception(e, request));
        }
    }

    @Route(path = "query/request")
    @Blocking
    public void query(RoutingExchange ex) {
        JsonRpcRequest request = ex.context().getBodyAsJson().mapTo(JsonRpcRequest.class);
        try {
            Object result = queryGateway.query(codec(request.params().get(0), getDomain(request.method())));
            ex.ok().end(success(result, request));
        } catch (Exception e) {
            ex.serverError().send(exception(e, request));
        }
    }

    @Route(path = "query/list")
    @Blocking
    public void queries(RoutingExchange ex) {
        JsonRpcRequest request = ex.context().getBodyAsJson().mapTo(JsonRpcRequest.class);
        try {
            Object result = queryGateway.queries(codec(request.params().get(0), getDomain(request.method())));
            ex.ok().end(success(result, request));
        } catch (Exception e) {
            ex.serverError().send(exception(e, request));
        }
    }

    public String success(Object payload, JsonRpcRequest jsonRpcRequest) {
        try {
            JsonRpcResponse jsonRpcResponse = new JsonRpcResponse(jsonRpcRequest.id(),
                                                                  objectMapper.writeValueAsString(payload));
            return objectMapper.writeValueAsString(jsonRpcResponse);
        } catch (Exception e) {
            throw new IllegalStateException(e.getMessage(), e);
        }
    }

    public String exception(Exception exception, JsonRpcRequest jsonRpcRequest) {
        try {
            JsonRpcResponse jsonRpcResponse = new JsonRpcResponse(jsonRpcRequest.id(), exception.getMessage(),
                                                                  exception.getClass().getCanonicalName());
            return objectMapper.writeValueAsString(jsonRpcResponse);
        } catch (Exception e) {
            throw new IllegalStateException(e.getMessage(), e);
        }
    }

    private Class<?> getDomain(String simpleName) {
        return registry.domainMaps.keySet().stream().filter(key -> key.endsWith(simpleName)).findFirst()
                                  .map(key -> registry.domainMaps.get(key))
                                  .orElseThrow(() -> new IllegalArgumentException("domain not found handler"));
    }


    private <T> Object codec(String msg, Class<T> clazz) {
        try {
            return objectMapper.readValue(msg, clazz);
        } catch (Exception e) {
            throw new IllegalStateException(e.getMessage(), e);
        }
    }

    public record JsonRpcRequest(int id, String method, List<String> params, String jsonrpc) {
        public JsonRpcRequest(int id, String method, List<String> params) {
            this(id, method, params, "2.0");
        }

    }

    public record JsonRpcResponse(int id, String result, String error, String exceptionType, String jsonrpc) {
        public JsonRpcResponse(int id, String result) {
            this(id, result, null, null, "2.0");
        }

        public JsonRpcResponse(int id, String error, String exceptionType) {
            this(id, null, error, exceptionType, "2.0");
        }

    }

}

File: server/src/main/kotlin/io/iamcyw/ams/ResourceManager.kt
package io.iamcyw.ams

import io.iamcyw.tower.commandhandling.gateway.CommandGateway
import io.iamcyw.tower.quarkus.runtime.DomainNameMappings
import io.iamcyw.tower.queryhandling.gateway.QueryGateway
import io.quarkus.vertx.web.Route
import io.quarkus.vertx.web.RoutingExchange
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import javax.enterprise.context.RequestScoped
import javax.inject.Inject

@RequestScoped
class ResourceManager @Inject constructor(
    private val commandGateway: CommandGateway,
    private val queryGateway: QueryGateway,
    private val registry: DomainNameMappings
) {

    @Route(path = "command/send")
    fun send(ex: RoutingExchange) {
        val request = ex.context().bodyAsJson.mapTo(JsonRpcRequest::class.java)
        try {
            val payload = code(request.params[0], registry.getDomain(request.method))
            commandGateway.send(payload)
            ex.ok().end()
        } catch (e: Exception) {
            ex.serverError().send(exception(e, request))
        }
    }

    @Route(path = "command/request")
    fun request(ex: RoutingExchange) {
        val request = ex.context().bodyAsJson.mapTo(JsonRpcRequest::class.java)
        try {
            val payload = code(request.params[0], registry.getDomain(request.method))
            val result = commandGateway.request<Any>(payload)
            ex.ok().send(success(result, request))
        } catch (e: Exception) {
            ex.serverError().send(exception(e, request))
        }
    }

    @Route(path = "query/*")
    fun query(ex: RoutingExchange) {
        val request = ex.context().bodyAsJson.mapTo(JsonRpcRequest::class.java)
        try {
            val payload = code(request.params[0], registry.getDomain(request.method))
            val result = queryGateway.query<Any>(payload)
            ex.ok().send(success(result, request))
        } catch (e: Exception) {
            ex.serverError().send(exception(e, request))
        }
    }

    private fun exception(ex: Exception, request: JsonRpcRequest): String {
        val response = JsonRpcResponse(request.id, exceptionType = ex::class.simpleName, error = ex.message)
        return Json.encodeToString(response)
    }

    private fun success(payload: Any?, request: JsonRpcRequest): String {
        val response = JsonRpcResponse(request.id, Json.encodeToString(payload))
        return Json.encodeToString(response)
    }

    @OptIn(InternalSerializationApi::class)
    fun <T : Any> code(msg: String, clazz: Class<T>) = Json.decodeFromString(clazz.kotlin.serializer(), msg)
}